### DISCUSSION TITLE
Be able to pre-build images for all sizes

### DISCUSSION THREAD

**Discussion by Sandstedt** [vercel/next.js#25040] [Ideas]
### Describe the feature you'd like to request

**Background:**
We have used next/image in a couple of live projects now, some really large sites with multiple languages and hundreds of products but maybe not that many visitors (but the ones who visit is very valuable).

**Problem:**
Our clients has complained about the images loads very slow sometimes, and that is when the image not already has been generated by next.js. We understand that this "generation on demand" is a feature to not store images not requested, and not d´inecrease build time. But if you have hundreds of images in 10 different sizes, it can takes weeks before all images is generated, cause a certain page is not visited that often, but when it does, we dont want the visitor to need to wait for all images to be generated.

And we notice the images doesn't seem to be running in parallel, so it only takes one image at a time.

### Describe the solution you'd like

Be able to have a build flag or option to generate all images in the build step.

### Describe alternatives you've considered

We could just beef up the node-server to be able to generate these images faster, but feels very unnecessary to have a big server standing mostly unproductive the rest of the time.

So another solution would be to force the generation from our side, by requesting all images externally in our CI with a GET request for each image and each size.

Or we could just bypass next/image generation and use a 3rd party service like Imgix, but as an extra cost.

**Comment by FabulousCupcake** [comment]
➕ 

Currently if you try to `next export` while using `next/image` you just get thrown to [this error page](https://nextjs.org/docs/messages/export-image-api).

It would be very cool if there is a fallback option so `next export` resizes/optimizes all the (non-external?) images from `<Image>` components during build time, and includes it in the Static HTML Export bundle.  

This would bring image optimization benefits on static webpages, which would be awesome.

**Reply by FabulousCupcake** [reply]
Looks like what I described is basically https://github.com/vercel/next.js/discussions/19065

**Reply by chemicalkosek** [reply]
I think OP doesn't want to use `next export` but `next start` with build time image generation.

**Reply by FabulousCupcake** [reply]
I see it now, you're right

**Reply by Sandstedt** [reply]
@FabulousCupcake That's also a really good usecase!

**Comment by chemicalkosek** [comment]
Even with the on-the-fly generation, the images are cached on the server for 60 seconds, right?
How did you manage to keep them longer? 
Do you have a nginx config or some other cache-control in next.config?
Can you share your nginx config for that or something? I'm not an expert in nginx. Thank you

Edit: I think I found something here: https://stackoverflow.com/questions/64972168/how-to-cache-nextjs-10-0-images-using-nginx
Either way, if you would share your solution, it would be great.

Relevant issue: https://github.com/vercel/next.js/issues/19914
Although that issue is about client side caching, it mentions the server side image cache too.
The image is generated again if 60 seconds have passed from last request.

**Reply by Sandstedt** [reply]
I'm not responsible for setting up our servers, so can't show an example, but we have set the cache time to 1 year. But should be easy to find in nqnix documentation I guess.
